### PR TITLE
move libevmcore to solidity

### PR DESCRIPTION
**Pull request by winsvega**
move instruction.cpp/h to from libethereum/libevmcore to solidity

DEPENDS:
{
"webthree-helpers":"evmcoresolidity",
"alethzero":"countsolidity"
}
connects to https://github.com/ethereum/webthree-umbrella/issues/410

### CODE REVIEW

**Comment by bobsummerwill** [comment]
OK - so you're trying a more minimal set of changes now?   And will keep iterating on this until it builds?
Good good.

**Comment by winsvega** [comment]
solidity builds fine already. need another pr to webthree-helpers.
but there is still one more dependence on libethereum in solidity test. 

alethzero depends on both solidity and libethereum thus it has redefenition problems
with evmcore constants

**Comment by bobsummerwill** [comment]
Yes, I don't think we can remove the dependencies from the tests.

**Comment by winsvega** [comment]
perhaps it is better to rename the constants inside solidity? then solidity tests and alethzero would not have redefinition problems

**Comment by bobsummerwill** [comment]
Yes, maybe just re-namespacing.

**Comment by winsvega** [comment]
just move it from dev::eth  to dev::solidity is fine?

**Comment by bobsummerwill** [comment]
That makes sense to me.

File: libsolidity/codegen/ArrayUtils.cpp [hunk 1]
```diff
@@ -56,7 +56,7 @@ void ArrayUtils::copyArrayToStorage(ArrayType const& _targetType, ArrayType cons
 	// stack: source_ref [source_length] target_ref
 	// store target_ref
 	for (unsigned i = _sourceType.sizeOnStack(); i > 0; --i)
-		m_context << eth::swapInstruction(i);
+		m_context << solidity::swapInstruction(i);
```
Comment (chriseth) [inline libsolidity/codegen/ArrayUtils.cpp]: I would guess that most of the prefixes are not necessary anymore.

File: libevmasm/Assembly.cpp [hunk 2]
```diff
@@ -121,7 +121,7 @@ ostream& Assembly::streamAsm(ostream& _out, string const& _prefix, StringMap con
 		switch (i.type())
 		{
 		case Operation:
-			_out << "  " << instructionInfo(i.instruction()).name  << "\t" << i.getJumpTypeAsString();
```
Comment (chriseth) [inline libevmasm/Assembly.cpp:124]: Why this change? It is against coding style.

File: libevmasm/CMakeLists.txt [hunk 3]
```diff
@@ -8,7 +8,7 @@ file(GLOB HEADERS "*.h")
 
 include_directories(BEFORE ..)
 add_library(${EXECUTABLE} ${SRC_LIST} ${HEADERS})
-eth_use(${EXECUTABLE} REQUIRED Eth::evmcore)
+eth_use(${EXECUTABLE} REQUIRED Dev::devcore Eth::ethcore)
```
Comment (chriseth) [inline libevmasm/CMakeLists.txt:11]: Didn't we want to remove ethcore?

Comment (winsvega) [inline libevmasm/CMakeLists.txt:11]: this was about evmcore dependency... 
now when I see this, we have to remove ethcore as well.  
in another PR maybe?

Comment (chriseth) [inline libevmasm/CMakeLists.txt:11]: Why did you _add_ ethcore here? evmcore does not depend on ethcore or does it?

Comment (winsvega) [inline libevmasm/CMakeLists.txt:11]: it seems that it does. I have errors if I remove this

**Comment by bobsummerwill** [comment]
This PR and the commits should be renamed.

The aim here is to duplicate the subset of evmasmcore used in solidity, to minimize our dependencies onto cpp-ethereum.     We'll still be stuck with a pretty meaty dependency ONLY in the tests, and it looks like we're going to have some dependencies onto devcore (typedefs, etc) remaining to be tackled over time, but they should be very stable and unchanging ones.

http://doublethinkco.github.io/cpp-ethereum-cross/images/dependency_graph.svg

So the new info here is that there appear to be `ethcore` dependencies in solidity **as well**?   Are those only in the solidity tests?    I would be surprised if we couldn't build Solidity without that dependency if `-DTESTS=Off` is passed on the command-line.   @winsvega?

**Comment by winsvega** [comment]
the main solidity evmcore is still depend on eth::ethcore
